k8s-build_tools/release.sh

```diff
@@ -0,0 +1,136 @@
+# We spin down the dormant application in order to save resources
+# There's an HPA in play, so we ensure we pre-heat the dormant app before we switch it over
+
+# We take the full tag to be the version. Currently that includes production_, and shouldn't,
+# but we're not going to hack around that edge case. It won't actually matter given the implementation below
+
+# This prototypoe is currently in deriv-com, but could be abstracted and injected into a build_tools/ for all our blue-green apps
+
+# Argument passing
+APP_NAME=$1
+VERSION=$2
+DEBUG=$3
+
+if [[ -n $DEBUG ]]; then 
+  set -x
+fi
+
+# Instead of leaving the other version running, we'll set it to dormant
+# This should make it easier to find. We'll also unset it's image
+DORMANT_VERSION=dormant
+
+# This is an arbitrary choice of 60s timeout. It can be changed if necessary.
+RELEASE_TIMEOUT=60s
+
+if [[ -z $APP_NAME || -z $VERSION ]]; then
+  echo "FATAL: Please pass APP_NAME and VERSION as arguments"
+  exit 1
+fi
+
+# Find what version is currently deployed
+CURRENT_VERSION=$(kubectl get -o jsonpath="{.spec.selector.version}" services/$APP_NAME)
+
+if [[ -z $CURRENT_VERSION ]]; then
+  echo "FATAL: Unable to find current version deployed"
+  exit 1
+fi
+
+# Find which deployment is live, using a label selector
+OLD_DEPLOYMENT=$(kubectl get deployment -l version=${CURRENT_VERSION},app=${APP_NAME} -o jsonpath='{.items[*].metadata.name}')
+
+if [[ -z $OLD_DEPLOYMENT ]]; then
+  echo "FATAL: Unable to find live deployment for $CURRENT_VERSION"
+  exit 1
+fi
+
+# Find which deployment is dormant
+NEW_DEPLOYMENT=$(kubectl get deployment -l version=${DORMANT_VERSION},app=${APP_NAME} -o jsonpath='{.items[*].metadata.name}')
+
+if [[ -z $NEW_DEPLOYMENT ]]; then
+  echo "FATAL: Unable to find spare dormant deployment"
+  exit 1
+fi
+
+if [[ $NEW_DEPLOYMENT == $OLD_DEPLOYMENT ]]; then
+  # This hasn't been observed, but just in case we end up with multiple labels somehow
+  echo "FATAL: Something is wrong, Dormant and Live are somehow the same"
+  exit 1
+fi
+
+# Need to know in advance what to scale the dormant deployment to, and apply that immediately after the image setting
+# We use the hpa because we don't want to get caught out when it was about to scale but hasn't yet
+# ( Could probably do this in one command, but I think it'd require eval'ing the output and we don't want to ever do that!
+DESIRED_REPLICAS=$(kubectl get hpa/${APP_NAME} -o jsonpath='{.status.desiredReplicas}')
+
+# But Desired replicas from the HPA can be < minimum, so we need to take that into account
+# Min Replicas over Current, in case there's an issue with the current deployment
+MIN_REPLICAS=$(kubectl get hpa/${APP_NAME} -o jsonpath='{.spec.minReplicas}')
+
+TARGET_REPLICAS=$(( DESIRED_REPLICAS > MIN_REPLICAS ? DESIRED_REPLICAS : MIN_REPLICAS ))
+
+# TODO - Setup the kubectl context so that kubectl works without arguments. We should not have to pass things like --server to it
```

Comment:
@afshinpaydar-binary Think this needs to move to the top as a note of what we need to do before running this.

* Set namespace in context
* Setup target kube server
* Auth credentials

All of that should be ideally tokenised in context, or worse case scenario in env vars ( But I don't think we resort to that ).

HTH

Reply:
Need to figure out how to do that!

Reply:
We can simply rely on `.kube/config` if we don’t want to pass parameters to this script, I guess